Gowin target

pepijndevos opened the pull request:
As the title suggests, it's a Gowin target for Nextpnr. It's a bit rough around the edges, but I just successfully ran a full flow and tested it on a GW1NR-9.

It's mostly a direct port from the old nexpnr-generic target, with added support for all LUTs, DFF reset/enable, correct validation checks, and pin constraints.

This means there is no deduplication and loading isn't instantaneous. When I add support for larger GW2A devices, I intend to revisit this target to add those things as needed. For the GW1N devices, it's not a real problem.

The main problem for now is that routing the attosoc demo takes forever. It has 53% utilization, but at about 8k unrouted nets it basically makes no progress at all.

For now the main things to do are

- [x] remove all debugging leftovers
- [x] figure out if attosoc routing congestion can be improved
- [x] address any feedback on this PR

### How to test

I have built this branch with the following CMake variables, where the notable thing is that the python interpreter is the same one as is used for Project Apicula. If you're using system Python, I guess these options are not needed.

```json
{
    "ARCH": "gowin",
    "BUILD_GUI": false,
    "PYTHON_EXECUTABLE": "/home/pepijn/code/apicula/env/bin/python",
    "GOWIN_BBA_EXECUTABLE": "/home/pepijn/code/apicula/env/bin/gowin_bba",
}
```

I'm pushing a new Apycula to Pypi as I write this, so you should be able to `pip install apycula` soon.

From there you should be able to use the latest yosys to `synth_gowin` an example program.
A simple way is to just `make blinky.json` in https://github.com/YosysHQ/apicula/tree/master/examples

Then you should be able to run nextpnr like so
```
nxtpnr-gowin --json apicula/examples/attosoc.json \
                --write out.json \
                --device GW1NR-9 \
                --family GW1N-9 \
                --package QN881 \
                --speed C6/I5 \
                --cst apicula/examples/gw1n9.cst
```

And then finally pack the bitstream with 

```
gowin_pack -d GW1N-9 -o blinky.json out.json
```

daveshah1 commented on gowin/arch.cc:
```diff
@@ -0,0 +1,1248 @@
+/*
+ *  nextpnr -- Next Generation Place and Route
+ *
+ *  Copyright (C) 2018  Clifford Wolf <clifford@symbioticeda.com>
```
Please update and add your own name

pepijndevos replied:
Should I do anything with @clairexen's entry? Seems to be using her deadname across the whole project except in _one_ location, so not sure what the accepted approach is for a new copy of old code.

daveshah1 replied:
Yes, update it in new code (afaik Yosys is mostly doing the same)

daveshah1 commented on gowin/arch.cc:
```diff
@@ -0,0 +1,1248 @@
+        col += segment;
+        break;
+    default:
+        snprintf(buf, 32, "R%dC%d_%s", row + 1, col + 1, wirename.c_str());
```
I'm not all that happy with snprintfs and maps of IdStrings, I'd really rather see a database with flat indices (like ice40), but I guess it's OK if  you are planning to move away from it

daveshah1 commented on gowin/arch.cc:
```diff
@@ -0,0 +1,1248 @@
+        archInfoToAttributes();
+        return retVal;
+    } else {
+        log_error("Generic architecture does not support placer '%s'\n", placer.c_str());
```
Gowin, not Generic

daveshah1 commented on gowin/arch.cc:
```diff
@@ -0,0 +1,1248 @@
+        if (ci->is_slice && ci->slice_clk != nullptr) {
+            Loc loc = getBelLocation(ci->bel);
+            int cls = loc.z / 2;
+            bool ff_used = ci->params.at(id_FF_USED).as_bool();
```
You could get a small speedup from adding ff_used to ArchCellInfo and assigning it in assignArchInfo

daveshah1 commented on gowin/arch_pybindings.cc:
```diff
@@ -0,0 +1,235 @@
+                    pass_through<float>>::def_wrap(ctx_cls, "addClock");
+
+    // Generic arch construction API
+    fn_wrapper_4a_v<Context, decltype(&Context::addWire), &Context::addWire, conv_from_str<IdString>,
```
All these add* pybindings aren't realistically needed, and can be removed (doing so will probably appreciably reduce build time and build memory usage)

daveshah1 commented on gowin/constids.inc:
```diff
@@ -0,0 +1,429 @@
+X(a_f)
+X(b_f)
+X(c_f)
+X(d_f)
```
Please add a newline at the end of file

daveshah1 commented on gowin/main.cc:
```diff
@@ -0,0 +1,82 @@
+po::options_description GowinCommandHandler::getArchOptions()
+{
+    po::options_description specific("Architecture specific options");
+    specific.add_options()("device", po::value<std::string>(), "device name");
```
I would like all new nextpnr arches to follow nextpnr-nexus and take a single device string like `GW1N-LV1QN48C6/I5` and extract family/package/speed from it. This will make integration with things like nMigen and LiteX easier.

daveshah1 commented on gowin/pack.cc:
```diff
@@ -0,0 +1,290 @@
+    for (auto cell : sorted(ctx->cells)) {
+        CellInfo *ci = cell.second;
+        if (ctx->verbose)
+            log_info("cell '%s' is of type '%s'\n", ci->name.c_str(ctx), ci->type.c_str(ctx));
```
tbf this probably needs refactoring in `generic` and other arches too, but it's neater to use `ctx->nameOf(ci)` here

daveshah1 commented on CMakeLists.txt:
```diff
@@ -107,9 +107,9 @@ if (MSVC)
     set(CMAKE_CXX_FLAGS_DEBUG "-Wall -fPIC -ggdb -pipe")
     if (USE_OPENMP)
-        set(CMAKE_CXX_FLAGS_RELEASE "-Wall -fPIC -O3 -g -pipe -fopenmp")
+        set(CMAKE_CXX_FLAGS_RELEASE "-Wall -fPIC -O0 -g -pipe -fopenmp")
```
Revert this back to O3

daveshah1 commented on gowin/arch.cc:
```diff
@@ -0,0 +1,1248 @@
+uint32_t Arch::getBelChecksum(BelId bel) const
+{
+    // FIXME
+    return 0;
```
returning the index here is fine

daveshah1 commented on gowin/arch.h:
```diff
@@ -0,0 +1,435 @@
+    delay_t estimateDelay(WireId src, WireId dst) const;
+    delay_t predictDelay(const NetInfo *net_info, const PortRef &sink) const;
+    delay_t getDelayEpsilon() const { return 0.001; }
+    delay_t getRipupDelayPenalty() const { return 0.015; }
```
I think you will probably get better results with a higher ripup penalty. For perspective ECP5 uses 400ps. epsilon might be worth increasing too.

daveshah1 commented on gowin/arch.cc:
```diff
@@ -0,0 +1,1248 @@
+    std::string line;
+    while (!in.eof()) {
+        std::getline(in, line);
+        if(!std::regex_match(line, match, iobre)) continue;
```
Can you add a warning if the regex fails but the line isn't all blank? An important constraint could be being missed

daveshah1 commented on gowin/arch.cc:
```diff
@@ -0,0 +1,1248 @@
+        // }
+        auto it = cells.find(net);
+        if (it == cells.end())
+            log_error("Cell %s not found", net.c_str(this));
```
Unless the Gowin tools behave like this, this should be a `continue` rather than an error - it's common for board constraint files to declare all the pins but only a subset are used in the design

daveshah1 commented on gowin/arch.cc:
```diff
@@ -0,0 +1,1248 @@
+    while (!in.eof()) {
+        std::getline(in, line);
+        if(!std::regex_match(line, match, iobre)) continue;
+        std::cout << match[1] << " " << match[2] << std::endl;
```
Remove this debugging, or replace it with a friendlier `log_info`

daveshah1 commented:
Thank you very much for submitting this. Please understand if some of my review comments appear harsh, as I am using a standard that is higher than the current quality of the nextpnr codebase - some of the things I have commented on I know just came from generic as-is - because improving things is always good :).

daveshah1 commented on gowin/arch.cc:
```diff
@@ -0,0 +1,1248 @@
+        tilePipDimZ[loc.x].resize(loc.y + 1);
+
+    gridDimX = std::max(gridDimX, loc.x + 1);
+    gridDimY = std::max(gridDimY, loc.x + 1);
```
loc.z

pepijndevos commented:
Makes complete sense. I'll probably get back to your feedback next week, because of a deadline coming Friday. I'm really excited to finally have a proper* nextpnr target, as a Christmas present to the community ;-)

* I agree the runtime loading of maps of IdString is not ideal...

daveshah1 commented on gowin/arch.cc:
```diff
@@ -0,0 +1,1248 @@
+// TODO represent wires more intelligently.
+IdString Arch::wireToGlobal(int &row, int &col, const DatabasePOD *db, IdString &wire)
+{
+    std::string wirename = wire.str(this);
```
`const std::string &` avoids a copy

daveshah1 commented on gowin/arch.cc:
```diff
@@ -0,0 +1,1248 @@
+        }
+        break;
+    }
+    if (len != id("")) {
```
I suspect `!= IdString()` will be faster

daveshah1 commented on gowin/arch.cc:
```diff
@@ -0,0 +1,1248 @@
+                belname = id(buf);
+                addBel(belname, id_SLICE, Loc(col, row, z), false);
+                snprintf(buf, 32, "R%dC%d_F%d", row + 1, col + 1, z);
+                wirename = id(buf);
```
is there any need for the wirename temporary variable - wouldn't doing `id(buf)` as an argument to `addBelOutput` etc be neater?

daveshah1 replied:
I also think you could have a list of pin names and do this in a loop

pepijndevos replied:
They are a mix of inputs and outputs, so I think it would be less clear in the end.

daveshah1 commented on gowin/arch.cc:
```diff
@@ -0,0 +1,1248 @@
+        } else {
+            ci->is_slice = false;
+        }
+        ci->user_group = int_or_default(ci->attrs, id("PACK_GROUP"), -1);
```
`PACK_GROUP` looks like dead code from `generic` to me

daveshah1 commented:
Great! A few other things I would suggest:

 - [ ] add a stub of the GUI files so that CI passes and you can do a `BUILD_GUI=yes` build combining gowin and other arches even if you don't care about the GUI for gowin ATM
 - [ ] add a brief note about Gowin to the README, even if just mentioning that it is highly experimental

daveshah1 commented on gowin/arch.cc:
```diff
@@ -484,8 +485,14 @@ void Arch::read_cst(std::istream &in) {
+            if(line.empty() || line.rfind("//", 0) == 0)
+                continue;
+            else
+                log_warning("Invalid constraint: %s", line.c_str());
```
Missing newline in message

daveshah1 commented on gowin/arch.cc:
```diff
@@ -496,8 +503,10 @@ void Arch::read_cst(std::istream &in) {
-        if (it == cells.end())
-            log_error("Cell %s not found", net.c_str(this));
+        if (it == cells.end()) {
+            log_info("Cell %s not found", net.c_str(this));
```
missing newline here too

daveshah1 commented on gowin/cells.cc:
```diff
@@ -83,14 +84,14 @@ void lut_to_lc(const Context *ctx, CellInfo *lut, CellInfo *lc, bool no_dff)
 
     if (no_dff) {
-        lc->params[id_FF_USED] = 0;
+        lc->ff_used = false;
```
Sorry, I may have explained this badly before. The correct way to do it is to set a param here and then set the ArchCellInfo field to mirror that param in `assignArchInfo` - ArchCellInfo should never be the only storage for important things, to avoid creating any more problems for loading/storing designs or post/pack simulations than needed (even if both these things are admittedly experimental right now)

pepijndevos replied:
FF_USED is not actually used on the Apicula side of things, so it's no important info. Actually the only place it's used is in validation that no DFF are used in the top two slices that don't have any.

daveshah1 replied:
And that would break if you saved the netlist between pack and place

pepijndevos commented:
Okay, I think I addressed all of your feedback now.

Routing attosoc is still.... not ideal. Increasing the ripup penalty did improve it slightly, now it slows down at 5k instead of 8k...

Also the build is still failing. Currently because Apicula requires Python 3.6 or higher, but of course the CI does not install Apicula at all.

pepijndevos commented:
Some things that need to be done still:

- [ ] add Apicula to CI, will require Python 3.6
- [ ] hunt down bug that breaks attosoc example. Issue with global clock routing I think, bug could be on the Apicula side.

optionally:
- [ ] better timing estimation
- [ ] dedicated global router
- [ ] more comprehensive constraints and configuration (clock domains?)

Anything else?

daveshah1 commented:
That list looks reasonable to me. I think making CI pass is the number one priority to get this merged which I'd rather do sooner than later, as GitHub doesn't charge for follow-up PRs.

smunaut commented:
Are you planning on squashing the commits for the merge ?
Or at least adding a `gowin: ` prefix to their title ?

pepijndevos commented:
Squashing them seems like a good plan, because I've also commited some WIP things that will be a pain if someone ever tries to bisect across these commits.

daveshah1 commented:
Now that CI is working, I think all that's left is squashing the commits if you want to do that; or I can merge-and squash.

An informal agreement to tidy up the arch database at some point would be nice too :)

pepijndevos commented:
Yea, feel free to merge-and-squash. Saves me some googling.

There will definitely be follow-up PRs to fix issues, clean up stuff, and add features.
I think once I add GW2A support going back and adding deduplication will be unavoidable.

daveshah1 commented:
Done :champagne: :confetti_ball:

whitequark commented:
Nice!
